ccp/src/cu/chunk.rs

```diff
@@ -0,0 +1,113 @@
+/*
+ * Copyright 2024 Fluence Labs Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+pub(crate) struct ChunkGenerator {
+    // inclusive
+    start: u64,
+    chunk_size: u64,
+    // exclusive
+    end: u64,
+}
+
+impl ChunkGenerator {
+    pub(crate) fn new(size: u64, chunk_size: u64) -> Self {
+        Self {
+            start: 0,
+            chunk_size,
+            end: size,
+        }
+    }
+}
+
+#[derive(Debug, PartialEq, Eq)]
+pub(crate) struct Chunk {
+    pub(crate) start: u64,
+    pub(crate) size: u64,
+}
+
+impl Iterator for ChunkGenerator {
+    type Item = Chunk;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        // we add 1 to substracted and 1 to result because we want to have None when end == start, not Some(0).
+        let remains = self.end.checked_sub(self.start + 1).map(|n| n + 1);
```

Comment:
I'd suggest to have matching here to early return in case if None to simplify the next line.

Reply:
and even better to use if with < or ==, in this case you don't need the logic with +1 and comment to explain it.